File: Raymond uber/Group_Isomorphic.cpp
/*
Given two strings s and t, determine if they are isomorphic.

Two strings are isomorphic if the characters in s can be replaced to get t.

All occurrences of a character must be replaced with another character while preserving the order of characters. 
No two characters may map to the same character but a character may map to itself.

For example,
Given "egg", "add", return true.

Given "foo", "bar", return false.

Given "paper", "title", return true.

Group them-> {"paper, title"}, {"egg","add"},{"hello","billy"}
*/
#include <iostream>
#include <string>
#include <vector>
#include <unordered_map>
using namespace std;
class Solution {
private:
	unordered_map<string, int>table;
public:
    vector<vector<string> > GroupIsomorphic(vector<string> & word) {
    	vector<vector<string> >res;
    	if(word.empty())return res;
        for(auto & str : word){
        	string tmp=convert(str);
        	if(!table.count(tmp)){
        		table[tmp]=res.size();
        		vector<string>v;
        		v.push_back(str);
        		res.push_back(v);
        	}
        	else res[table[tmp]].push_back(str);

        }
        return res;
    }
    string convert(string str){
        unordered_map<char, int>map;
        if(str.length() <=1 )return "#";
        string res="";
        int diff=0;
        for(auto ch : str){
            if(!map.count(ch))map[ch]=diff++;
            res+=to_string(map[ch])+'#';
        }
        return res;

    }
};
int main(){
	Solution ss;
	string b[]={"egg","add","laa","title","paper","hello","billy","kitten","how","today"};
	vector<string> word(b, b + sizeof(b) / sizeof(b[0]));
	vector<vector<string> >res= ss.GroupIsomorphic(word);

	for(auto & data : res){
		for(auto & s : data)cout << s << " ";
		putchar('\n');
	}
	return 0;
}

File: Raymond uber/Group_Isomorphic.java
import java.io.*;
import java.util.*;

/*
Given two strings s and t, determine if they are isomorphic.

Two strings are isomorphic if the characters in s can be replaced to get t.

All occurrences of a character must be replaced with another character while preserving the order of characters. 
No two characters may map to the same character but a character may map to itself.

For example,
Given "egg", "add", return true.

Given "foo", "bar", return false.

Given "paper", "title", return true.

Group them-> {"paper, title"}, {"egg","add"},{"hello","billy"}
*/

public class Solution {
  
  public static List<List<String>> Group(String[] strs) {
    List<List<String>> ret = new ArrayList<>();
    HashMap<String, List<String>> map = new HashMap<>();
    for(String s : strs) {
      String key = hash(s);
      if(!map.containsKey(key)) {
        map.put(key,new ArrayList<String>());
      }
        map.get(key).add(s);
    }
    for(String key : map.keySet()) {
      ret.add(map.get(key));
    }
    return ret;
    
}
  
  public static String hash(String str) {
    HashMap<Character, Integer > map = new HashMap<>();
    StringBuilder sb = new StringBuilder();
    int cnt = 0;
    for(int i = 0;i<str.length();i++) {
      char c =str.charAt(i);
      if(!map.containsKey(c)) {
        sb.append(cnt+"#");
        map.put(c,cnt);
        cnt++;
      } else {
        sb.append(map.get(c)+"#");
      }
    }
    return sb.toString();
  }

  public static void main(String[] args) {
    String[] b ={"egg","add","laa","title","paper","hello","billy","kitten","how","today"};
    for(List<String> list : Group(b)) {
      for(String each : list) {
        System.out.print(each + " ");
      }
      System.out.println();
    }
    
  }
  
  
}
